Translate tp1 commands to tp2

donnyyung opened the pull request:
Here's my PR for translating tp1 to tp2 commands. It's not exhaustive, but it tries to capture most of the commands (--swap-deployment, --run, --docker-run), as well as the various flags that can be used with those commands.  

---
## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes. - No docs PR for this needed. I do want to make a doc of tp1 -> tp2 but that's a separate work item
 - [x] My change is adequately tested.
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

Commit:
Version 0 translating some legacy commands to tp2

This is the first attempt at translating the swap-deployment commands
from telepresence 1 into telepresence 2.  It currently just spits out
the telepresence 2 command out to the user.  In future commits, I will
try to automatically run said command, as well as include more commands
from telepresence 1.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add legacy_command tests

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Better return msgs to user about tp1->tp2 changes

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Added new commands + slight refactor

Before I only translated --swap-deployment, I've now added translations
for --run and --run-shell, and some additional tests for those

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Make telepresence --help work again

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Automatically run translated commands

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
bump changelog

Signed-off-by: Donny Yung <donaldyung@datawire.io>

donnyyung commented on pkg/client/cli/command.go:
```diff
@@ -96,34 +107,41 @@ func Command(ctx context.Context) *cobra.Command {
+		Short:              myName,
+		Long:               help,
+		RunE:               RunSubcommands,
+		PostRunE:           checkLegacyCmd,
```
I added it as a PostRun so that this logic didn't get gross (in case we want to remove this at some point in the future).  But that being said, when users use telepresence 1 commands, the help text will be printed to the terminal (from RunSubCommands) and then my command will run. 

Example output:
```
...
  other Telepresence flags:
      --no-report   turn off anonymous crash reports and log submission on failure

Use "telepresence [command] --help" for more information about a command.

You used a telepresence 1 command that roughly translates to the following:
telepresence intercept echo-easy --port 9090 -- sleep 1
running...
Launching Telepresence Daemon v2.2.2-14-g56008474-1620355231 (api v3)
Connecting to traffic manager...
...
```
I considered this to not be a negative since I think while we want tp1 commands to work, we do want people to eventually migrate to using tp2 commands so the help text will maybe push them to explore.  But if you think it's messy, I can probably merge it into the RunSubcommands function.

Commit:
run make generate

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Have scout report legacy telepresence usage

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Slight output changes

I was preparing for my demo and didn't like the way the output was
formatted so I made some changes

Signed-off-by: Donny Yung <donaldyung@datawire.io>

thallgren commented on pkg/client/cli/legacy_command.go:
```diff
@@ -218,12 +218,15 @@ func checkLegacyCmd(cmd *cobra.Command, args []string) error {
+	// Separate the help output from the tp1 -> tp2 translation
 	if tp2Cmd != "" {
-		fmt.Fprintf(cmd.OutOrStderr(), "\nYou used a telepresence 1 command that roughly translates to the following:\ntelepresence %s\n", tp2Cmd)
+		fmt.Fprintf(cmd.OutOrStderr(), "\nLegacy telepresence command used\n")
```
This is really cool. Especially this part:
```
Legacy telepresence command used
Command roughly translates to the following in Telepresence 2:
telepresence intercept echo-easy --port 9090 -- python3 -m http.server 9090
running...
```
but does it have to be prefixed by the whole usage-blob? In my (very personal opinion), that makes the message less obvious.

donnyyung replied:
Yeah I'm going to take the help text out. I thought it was _maybe_ useful for pointing people at the help info so they could learn the new syntax w/o crutching on the legacy Telepresence commands for too long. 

But a few people have mentioned not liking it and it obscures the actual message we want. I might add to the text above something like: "For more information about the new commands, use telepresence --help" to point people to the help text to learn more. What do you think about that? Or do you think `--help` is obvious enough that we don't need to draw attention to it. Kenn also suggested I ask Brian his opinion so I'll do that in parallel and add a comment with what he suggested.

--- edit ---
Brian agreed with this sentiment so I have removed the help text from showing when running a legacy Telepresence command.

thallgren commented on pkg/client/cli/legacy_command_test.go:
```diff
@@ -24,13 +24,13 @@ func Test_legacyCommands(t *testing.T) {
 			inputLegacyCommand: "telepresence --swap-deployment myserver --method inject-tcp --expose 9090 --run python3 -m http.server 9090",
 			outputTP2Command:   "intercept myserver --port 9090 -- python3 -m http.server 9090",
-			msg:                "Telepresence 2 doesn't have methods. You can use --docker-run for container, otherwise tp2 works similarly to vpn-tcp",
+			msg:                "Telepresence 2 doesn't have methods. You can use --docker-run for container, otherwise it works similarly to vpn-tcp\n",
```
Suggest using the term "proxying methods" instead of just "methods" since that term is very generic and somewhat close to "mechanisms".

thallgren reviewed (approved):
Looks great! I made a couple of suggestions. Feel free to ignore.

Commit:
Address PR feedback

Included some slight changes to the output and also not spitting out the
help text when running a legacy Telepresence command

Signed-off-by: Donny Yung <donaldyung@datawire.io>
